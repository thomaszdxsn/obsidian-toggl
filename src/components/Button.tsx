import React from 'react'
import { ClipLoader } from 'react-spinners'

interface Props extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  children: React.ReactNode
  isLoading?: boolean
  loadingIconSize?: number
}


export const Button = ({ children, isLoading, loadingIconSize=13, ...props }: Props) => {
  return <button
    {...props}
    disabled={isLoading || props.disabled}
  >{isLoading ?
    <ClipLoader size={loadingIconSize}
      cssOverride={{
        borderColor: "var(--accent-h)",
      }}
    />
    : children}</button>
}