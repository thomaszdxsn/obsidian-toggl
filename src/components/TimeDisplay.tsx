import React from 'react'
import clsx from 'clsx'
import { css } from '@emotion/css'

interface Props {
  className?: string
  children: React.ReactNode
}

export const TimeDisplay = ({ children, className }: Props) => {
  return (
    <div className={
      clsx(css`
        font-variant-numeric: tabular-nums;
        `, className)
    }>
      {children}
    </div>
  )
}