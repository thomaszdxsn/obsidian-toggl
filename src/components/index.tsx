export { CurrentTimer } from './CurrentTimer'
export { MainView } from './MainView'
export { Provider } from './Provider'
export { TimerCard } from './TimerCard'
export { TimerList } from './TimerList'
export { TimerForm } from './TimerForm'
export { Button } from './Button'
export { Timeline } from './Timeline'