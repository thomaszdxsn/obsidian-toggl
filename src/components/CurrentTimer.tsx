import React from "react";
import { useAtom, useAtomValue, useSetAtom } from "jotai";
import { useStopTimerMutation } from "../hooks";
import {
  currentEntryAtom,
  currentEntryProjectAtom,
  currentTimerAtom,
  passedSecondsAtom,
  passedTimeAtom,
} from "../atoms";
import { css } from "@emotion/css";
import { FiPause } from "react-icons/fi";
import { TimeDisplay } from "./TimeDisplay";
import { Button } from "./Button";
import { generateTimerId } from "../utils";

export const CurrentTimer = ({
  placeholder = "No time entry running",
}: {
  placeholder?: string;
}) => {
  const passed = useAtomValue(passedTimeAtom);
  const setPassedNumbers = useSetAtom(passedSecondsAtom);
  const [currentEntry, setCurrentEntry] = useAtom(currentEntryAtom);
  const currentTimer = useAtomValue(currentTimerAtom);
  const stopTimerMutation = useStopTimerMutation(
    currentTimer ? generateTimerId(currentTimer) : undefined,
  );
  const currentProject = useAtomValue(currentEntryProjectAtom);
  const onStop = () => {
    if (!currentEntry) {
      return;
    }
    stopTimerMutation.mutate(
      { timeEntryId: currentEntry.id, workspaceId: currentEntry.workspace_id },
      {
        onSuccess: () => {
          setCurrentEntry(null);
          setPassedNumbers(null);
        },
      },
    );
  };
  if (!passed) {
    return (
      <div
        className={css`
          height: 100px;
          display: flex;
          justify-content: center;
          align-items: center;
          border-radius: var(--radius-m);
          background-color: var(--background-primary);
        `}
      >
        {placeholder}
      </div>
    );
  }
  return (
    <div
      className={css`
        display: flex;
        flex-direction: column;
        background-color: var(--background-primary);
        padding: var(--size-4-4);
        border-radius: var(--radius-m);
        height: 100px;
        container-type: inline-size;
      `}
    >
      <Button
        onClick={onStop}
        isLoading={stopTimerMutation.isPending}
        className={css`
          flex: 1;
          cursor: ${stopTimerMutation.isPending ? "wait" : "pointer"};
          margin-bottom: var(--size-4-2);
        `}
      >
        <FiPause />
      </Button>
      <div
        className={css`
          display: flex;
          justify-content: center;
          align-items: center;
          flex-wrap: wrap;
          @container (min-width: 200px) {
            justify-content: space-between;
          }
        `}
      >
        <div>
          {currentProject && (
            <div style={{ color: currentProject.color }}>
              {currentProject.name}
            </div>
          )}
          {currentEntry && (
            <div
              className={css`
                font-size: var(--font-smallest);
                color: var(--text-muted);
              `}
            >
              {currentEntry.description}
            </div>
          )}
        </div>
        <TimeDisplay
          className={css`
            font-size: var(--font-ui-large);
            font-weight: var(--font-semibold);
          `}
        >
          {passed}
        </TimeDisplay>
      </div>
    </div>
  );
};
