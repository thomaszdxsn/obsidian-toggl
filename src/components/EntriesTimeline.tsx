import React, { useMemo } from "react";
import { Project, TimeEntry } from "../interfaces";
import { Timeline } from "./Timeline";
import { calcPercentage, formatSeconds, formatTime } from "../utils";
import { css } from "@emotion/css";
import clsx from "clsx";
import { TimeDisplay } from "./TimeDisplay";

interface Props {
  className?: string;
  entries: TimeEntry[];
  projectDict: Record<string, Project>;
  entryMinHeight?: number;
  equalEntryHeight?: boolean;
}

export const EntriesTimeline = ({
  className,
  entries,
  projectDict,
  entryMinHeight = 100,
  equalEntryHeight = true,
}: Props) => {
  const items = useMemo(() => {
    const totalDuration = entries.reduce((acc, entry) => {
      return acc + entry.duration;
    }, 0);
    return entries
      .slice()
      .sort((a, b) => (a.start < b.start ? 1 : -1))
      .map((entry) => {
        const project = entry.project_id
          ? projectDict[entry.project_id]
          : undefined;
        const percentage = equalEntryHeight
          ? calcPercentage(1, entries.length)
          : calcPercentage(entry.duration, totalDuration);
        const label = (
          <div
            className={css`
              display: flex;
              justify-content: space-between;
              height: 100%;
              gap: 8px;
              padding-right: 16px;
            `}
          >
            <div
              className={css`
                flex-shrink: 0;
                display: flex;
                height: 100%;
                flex-direction: column;
                justify-content: space-between;
                min-width: 0;
                padding: 16px 0;
                white-space: nowrap;
                overflow: hidden;
                text-overflow: ellipsis;
              `}
            >
              <div style={{ color: project?.color }}>{project?.name}</div>
              <div
                className={css`
                  white-space: nowrap;
                  overflow: hidden;
                  text-overflow: ellipsis;
                `}
              >
                {entry.description}
              </div>
              <div
                className={css`
                  color: var(--text-accent);
                  white-space: nowrap;
                  overflow: hidden;
                  text-overflow: ellipsis;
                `}
              >
                üè∑Ô∏è {entry.tags.join(",")}
              </div>
            </div>
            <div
              className={css`
                height: 100%;
                display: flex;
                flex-direction: column;
                justify-content: center;
                align-items: end;
                gap: 8px;
                flex-shrink: 0;
              `}
            >
              <TimeDisplay
                className={css`
                  font-size: 1rem;
                  font-weight: var(--font-semibold);
                `}
              >
                {formatSeconds(entry.duration)}
              </TimeDisplay>
              <TimeDisplay
                className={css`
                  font-size: 0.75rem;
                `}
              >
                {formatTime(entry.start)} - {formatTime(entry.stop)}
              </TimeDisplay>
            </div>
          </div>
        );
        return {
          color: project?.color || "var(--background-primary)",
          label: label,
          percentage: percentage,
        };
      });
  }, [entries, projectDict]);
  const totalHeight = useMemo(() => {
    const smallestPercentage = Math.min(
      ...items.map((item) => item.percentage),
    );
    return items.reduce((acc, item) => {
      return acc + (item.percentage / smallestPercentage) * entryMinHeight;
    }, 0);
  }, [items, entryMinHeight]);
  return (
    <Timeline
      direction="vertical"
      className={clsx(
        className,
        css`
          height: ${totalHeight}px;
        `,
      )}
      items={items}
      size="medium"
      gap={1}
      finished
    />
  );
};
