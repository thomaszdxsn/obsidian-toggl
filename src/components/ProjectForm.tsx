import React from 'react'
import { useAtomValue } from "jotai";
import { useMemo } from "react";
import { useForm } from "react-hook-form";
import { projectsAtom } from "src/atoms";
import { Button } from './Button';
import { FiCheck, FiCornerUpLeft, FiRefreshCw } from 'react-icons/fi';
import { css } from '@emotion/css';
import { useCreateProjectMutation } from 'src/hooks';

interface Props {
  onSucecss: () => void
  projectId?: number
}

const BUILTIN_COLORS = [
  '#0b83d9',
  '#9e5bd9',
  '#d94182',
  '#e36a00',
  '#bf7000',
  '#2da608',
  '#06a893',
  '#c9806b',
  '#465bb3',
  '#990099',
  '#c7af14',
  '#566614',
  '#d92b2b',
  '#525266'
]

export const ProjectForm = ({ onSucecss, projectId }: Props) => {
  const projects = useAtomValue(projectsAtom)
  const project = useMemo(() => projects.find(project => project.id === projectId), [projectId])
  const form = useForm({
    defaultValues: {
      name: project?.name || "",
      color: project?.color || BUILTIN_COLORS[0]
    },
  })
  const createProjectMutation = useCreateProjectMutation()

  const onSubmit = async () => {
    const data = form.getValues()
    // TODO: distinct create and update
    createProjectMutation.mutate(data, {
      onSuccess: () => onSucecss?.()
    })
  }

  const isPending = createProjectMutation.isPending

  return (
    <form
      className={css`
    display: flex;
    flex-direction: column;
    gap: var(--size-4-2);
  `}
      onSubmit={form.handleSubmit(onSubmit)}
    >
      <label
        className={css`
                display: flex;
                flex-direction: column;
                gap: var(--size-4-1);
              `}
      >
        <span>Name</span>
        <input type="text" {...form.register("name")} />
      </label>
      <label
        className={css`
                display: flex;
                flex-direction: column;
                gap: var(--size-4-1);
              `}
      >
        <span>Color</span>
        <div className={css`
          display: flex;
          gap: var(--size-4-1);
        `}>
          {BUILTIN_COLORS.map(color => {
            return <label key={color} htmlFor={color}>
              <input
                type="radio"
                id={color}
                value={color}
                className={css`display: none;`}
                {...form.register("color")}
              />
              <div
                className={css`
                  display: inline-flex;
                  align-items: center;
                  justify-content: center;
                  width: var(--size-4-4);
                  height: var(--size-4-4);
                  border-radius: 50%;
                  background-color: ${color};
                `}
              >
                {form.watch("color") === color && <FiCheck color="white" />}
              </div>
            </label>
          })}
        </div>
      </label>
      <footer className={
        css`
          display: flex;
          gap: var(--size-4-2);
          justify-content: center;
        `
      }>
        <Button
          type="submit"
          disabled={isPending}
          className={css`
            cursor: pointer;
          `}
        >
          {isPending ?
            <FiRefreshCw className={
              css`
                    animation: spin 1s linear infinite
                  `
            } />
            : <FiCornerUpLeft />}
        </Button>
      </footer>
    </form>
  )
}