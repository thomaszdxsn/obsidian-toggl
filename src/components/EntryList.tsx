import React from "react";
import { useAtomValue } from "jotai";
import { projectDictAtom, todayTimeEntriesAtom } from "src/atoms";
import { EntriesTimeline } from "./EntriesTimeline";
import { isActiveEntry } from "src/utils";
import { css } from "@emotion/css";

export const EntryList = () => {
  const todayEntries = useAtomValue(todayTimeEntriesAtom);
  const entries = todayEntries.filter((entry) => !isActiveEntry(entry));
  const projectDict = useAtomValue(projectDictAtom);
  console.log({ entries });
  return (
    <div
      className={css`
        max-height: 30vh;
        overflow-y: auto;
      `}
    >
      <EntriesTimeline entries={entries} projectDict={projectDict} />
    </div>
  );
};
