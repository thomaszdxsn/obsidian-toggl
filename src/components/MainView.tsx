import { css } from "@emotion/css"
import React, { useMemo } from "react"
import { usePlugin } from "../hooks"
import { ProjectModal, TimerDetailModal } from "../plugin/modal"
import { CurrentTimer } from "./CurrentTimer"
import { TimerList } from "./TimerList"
import { FiCornerUpLeft, FiFolder, FiPlus } from 'react-icons/fi'
import { Button } from "./Button"
import { useAtomValue, useSetAtom } from "jotai"
import { activeProjectsAtom, projectDictAtom, todayTimeEntriesAtom, viewAtom } from "src/atoms"
import dayjs from "dayjs"
import { Timeline } from "./Timeline"
import { RefreshButton } from "./RefreshButton"
import { EntryList } from "./EntryList"
import { ProjectList } from "./ProjectList"

export const MainView = () => {
  const currentView = useAtomValue(viewAtom)
  const viewContent = useMemo(() => {
    switch (currentView) {
      case "homepage":
        return <HomePageView />
      case "projects":
        return <ProjectsView />
      case "tags":
        return <TagsView />
      default:
        return null
    }
  }, [currentView])
  return (
    <div
      className={css`
				display: flex;
				flex-direction: column;
				gap: var(--size-4-2);
			`}
    >
      <TodayTimeline />
      <CurrentTimer />
      <section className={css`
				display: flex;
				flex-direction: column;
				gap: var(--size-4-2);
			`}>
        {viewContent}
      </section>
    </div>
  )
}


export const TodayTimeline = () => {
  const entries = useAtomValue(todayTimeEntriesAtom)
  const projectDict = useAtomValue(projectDictAtom)
  const timelineItems = useMemo(() => {
    const total = entries.reduce((acc, entry) => {
      const duration = entry.duration !== -1 ? entry.duration : dayjs().diff(entry.start, "second")
      return acc + duration
    }, 0)
    return entries.sort((a, b) => a.start < b.start ? - 1 : 1).map(entry => {
      const project = entry.project_id ? projectDict[entry.project_id] : null
      const duration = entry.duration !== -1 ? entry.duration : dayjs().diff(entry.start, "second")
      const percentage = duration / total
      return {
        color: project?.color || "var(--background-primary)",
        label: project?.name,
        percentage: percentage
      }
    })

  }, [entries, projectDict])
  return (
    <div className={css`
        background: var(--background-primary);
        border-start-end-radius: 1rem;
        border-end-end-radius: 1rem;
        flex: 1;
      `}>
      <Timeline items={timelineItems} direction="horizontal" showLabel={false} />
    </div>
  )
}


export const HomePageView = () => {
  const plugin = usePlugin()
  const app = plugin.app
  const switchView = useSetAtom(viewAtom)
  const onAddTimer = () => {
    const modal = new TimerDetailModal(app, plugin)
    modal.open()
  }
  const onSwitchToProjectsView = () => {
    switchView("projects")
  }
  return (
    <>
      <div className={css`
          display: flex;
          gap: var(--size-4-2);
          justify-content: flex-end;
        `}>
        <Button onClick={onSwitchToProjectsView} title="projects">
          <FiFolder />
        </Button>
        <RefreshButton />
        <Button onClick={onAddTimer} title="add new timer">
          <FiPlus />
        </Button>
      </div>
      <TimerList plugin={plugin} />
      <EntryList />
    </>
  )
}


export const ProjectsView = () => {
  const plugin = usePlugin()
  const switchView = useSetAtom(viewAtom)
  const backToHomePage = () => switchView("homepage")
  const onAddProject = () => {
    const modal = new ProjectModal(plugin)
    modal.open()
  }
  const projects = useAtomValue(activeProjectsAtom)
  return (
    <>
      <div className={css`
      display: flex;
      justify-content: space-between;
    `}>
        <Button onClick={backToHomePage} title="projects">
          <FiCornerUpLeft />
        </Button>
        <div className={css`
          display: flex;
          gap: var(--size-4-2);
        `}>
          <Button onClick={onAddProject} title="add new projeect">
            <FiPlus />
          </Button>
        </div>
      </div>
      <ProjectList data={projects} />
    </>
  )
}


export const TagsView = () => {
  return null
}