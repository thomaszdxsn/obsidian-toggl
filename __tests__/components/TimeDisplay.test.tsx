import { TimeDisplay } from "../../src/components/TimeDisplay";
import "@testing-library/jest-dom"
import React from 'react'
import {render}from '@testing-library/react'


it("render correctly", () => {
  const {getByText, container} = render(<TimeDisplay>10:00</TimeDisplay>)
  expect(getByText("10:00")).toBeInTheDocument()
  expect(container).toMatchSnapshot()
})