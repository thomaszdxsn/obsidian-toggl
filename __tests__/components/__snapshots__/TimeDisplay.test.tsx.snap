// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<div>
  <div
    class="css-4dyp2g"
  >
    10:00
  </div>
</div>
`;
